QuarkusIntegrationTest cannot be run with failsafe in JVM mode

## Describe the bug

When trying to use failsafe to run a `@QuarkusIntegrationTest` in JVM mode (outside the `native` profile), test executions fails.

### Expected behavior

As documented [here](https://quarkus.io/guides/getting-started-testing#quarkus-integration-test), QuarkusIntegrationTest can be run with failsafe.

### Actual behavior

Test execution fails with the following error message, `GreetingResourceTestIT` being the test class annotated with `@QuarkusIntegrationTest`:

```
[INFO] Running org.acme.rest.GreetingResourceIT
2021-07-08 18:50:04,290 WARN  [io.qua.hib.orm.run.pro.ProxyDefinitions] (main) Could not generate an enhanced proxy for entity 'org.acme.rest.FooEntity' (class='org.acme.rest.FooEntity') as it's final. Your application might perform better if we're allowed to extend it.
2021-07-08 18:50:04,942 INFO  [io.quarkus] (main) Quarkus 2.0.1.Final on JVM started in 2.115s. Listening on: http://localhost:8081
2021-07-08 18:50:04,942 INFO  [io.quarkus] (main) Profile test activated. 
2021-07-08 18:50:04,942 INFO  [io.quarkus] (main) Installed features: [agroal, cdi, hibernate-orm, hibernate-orm-panache-kotlin, jdbc-h2, kotlin, narayana-jta, resteasy, resteasy-jackson, smallrye-context-propagation]
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 4.254 s <<< FAILURE! - in org.acme.rest.GreetingResourceIT
[ERROR] org.acme.rest.GreetingResourceIT  Time elapsed: 4.254 s  <<< ERROR!
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
Caused by: java.lang.reflect.InvocationTargetException
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: No bean found for required type [class org.acme.rest.GreetingResourceIT] and qualifiers [[]]
```

## To Reproduce

1. Clone [the reproduction repository](https://github.com/languitar/quarkus-intergration-test-failsafe-issue)
2. `mvn verify`

I also tried reproducing this from an even smaller project, but then this bug is not triggered. So there might be an interaction with other extensions configured in this project.

### Configuration

included

## Environment (please complete the following information):

### Output of `uname -a` or `ver`

```
Linux pres 5.10.47-1-lts #1 SMP Wed, 30 Jun 2021 13:52:19 +0000 x86_64 GNU/Linux
```


### Output of `java -version`

```
openjdk version "11.0.11" 2021-04-20
OpenJDK Runtime Environment (build 11.0.11+9)
OpenJDK 64-Bit Server VM (build 11.0.11+9, mixed mode)
```

### GraalVM version (if different from Java)

not needed

### Quarkus version or git rev

2.0.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

```
Apache Maven 3.8.1 (NON-CANONICAL_2021-04-26T21:52:54Z_root)
Maven home: /opt/maven
Java version: 11.0.11, vendor: Oracle Corporation, runtime: /usr/lib/jvm/java-11-openjdk
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.10.47-1-lts", arch: "amd64", family: "unix"
```

## Additional context
(Add any other context about the problem here.)

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusIntegrationTestExtension.java
@@ -186,6 +186,7 @@ public class QuarkusIntegrationTestExtension
 
     @Override
     public void postProcessTestInstance(Object testInstance, ExtensionContext context) {
+        ensureStarted(context);
         if (!failedBoot) {
             doProcessTestInstance(testInstance, context);
         }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -485,7 +485,7 @@ public class QuarkusTestExtension
 
     @Override
     public void beforeEach(ExtensionContext context) throws Exception {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(context.getRequiredTestClass())) {
             return;
         }
         resetHangTimeout();
@@ -554,7 +554,7 @@ public class QuarkusTestExtension
 
     @Override
     public void afterEach(ExtensionContext context) throws Exception {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(context.getRequiredTestClass())) {
             return;
         }
         resetHangTimeout();
@@ -618,11 +618,14 @@ public class QuarkusTestExtension
                 constructor.newInstance(actualTestInstance, actualTestMethod));
     }
 
-    private boolean isNativeOrIntegrationTest() {
+    private boolean isNativeOrIntegrationTest(Class<?> clazz) {
         for (Class<?> i : currentTestClassStack) {
             if (i.isAnnotationPresent(NativeImageTest.class) || i.isAnnotationPresent(QuarkusIntegrationTest.class)) {
                 return true;
             }
+        }
+        if (clazz.isAnnotationPresent(NativeImageTest.class) || clazz.isAnnotationPresent(QuarkusIntegrationTest.class)) {
+            return true;
         }
         return false;
     }
@@ -691,7 +694,7 @@ public class QuarkusTestExtension
         currentTestClassStack.push(context.getRequiredTestClass());
         //set the right launch mode in the outer CL, used by the HTTP host config source
         ProfileManager.setLaunchMode(LaunchMode.TEST);
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(context.getRequiredTestClass())) {
             return;
         }
         resetHangTimeout();
@@ -730,7 +733,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptBeforeAllMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -747,7 +750,7 @@ public class QuarkusTestExtension
     @Override
     public <T> T interceptTestClassConstructor(Invocation<T> invocation,
             ReflectiveInvocationContext<Constructor<T>> invocationContext, ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             return invocation.proceed();
         }
         resetHangTimeout();
@@ -841,7 +844,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptBeforeEachMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -852,7 +855,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -863,7 +866,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptTestTemplateMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -875,7 +878,7 @@ public class QuarkusTestExtension
     @Override
     public <T> T interceptTestFactoryMethod(Invocation<T> invocation,
             ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             return invocation.proceed();
         }
         T result = (T) runExtensionMethod(invocationContext, extensionContext);
@@ -886,7 +889,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptAfterEachMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -897,7 +900,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptAfterAllMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeOrIntegrationTest()) {
+        if (isNativeOrIntegrationTest(extensionContext.getRequiredTestClass())) {
             invocation.proceed();
             return;
         }
@@ -1045,7 +1048,7 @@ public class QuarkusTestExtension
     public void afterAll(ExtensionContext context) throws Exception {
         resetHangTimeout();
         try {
-            if (!isNativeOrIntegrationTest() && (runningQuarkusApplication != null)) {
+            if (!isNativeOrIntegrationTest(context.getRequiredTestClass()) && (runningQuarkusApplication != null)) {
                 popMockContext();
             }
             if (originalCl != null) {

